### ISSUE TITLE
Some aggregations and functions produce incorrect results for native histograms

### ISSUE THREAD

**Issue by NeerajGartia21** [prometheus/prometheus#13934]
Currently there are some operators and aggregations that produce either incorrect or inconsistent results for native histograms in some cases. Which are-

|  | Args | Current Behaviour | Expected Behaviour                | Status |
| :-------- | :------- | :------------------------- | :--------- | :------- |
| `+` and `-` | `float`, `histogram` in any order | returns `float` | no result, add info annotation | <li>- [x] </li>  |
| `*` and `/` | `histogram`, `histogram` | returns `float` | no result, add info annotation | <li>- [x] </li> |
| `/` | `float`, `histogram` (in this order! `histogram`, `float` is fine) | returns `float` ignoring histogram | no result, add info annotation | <li>- [x] </li> |
| `/` | `histogram`, `0` | returns `histogram` with every value divided by zero | return `histogram` with `sum` and `count` as `NaN` or `Inf` and all buckets removed | <li>- [x] </li> |
| `min`/`max`/`quantile` | combination of `histogram` and `float` | returns `scalar` by taking `histogram` as 0 | return `min`/`max`/`quantile` of all the floats and add info annotation |<li>- [x] </li> |
| `min`/`max`/`quantile` | only `histogram` | returns `scalar` by taking `histogram` as 0 | no result, add info annotation | <li>- [x] </li> |
| `stdvar`, `stddev` | combination of `histogram` and `float` | ignores histogram and only processes floats | no change in result but add info annotation | <li>- [x] </li> |
| `stdvar`, `stddev`| only `histogram` | no result | no result, add info annotation | <li>- [x] </li> |
|`==` and `!=`  without `bool` modifier| `float`, `histogram` in any order | treat `histogram` as 0, return either histogram or `float` depending on order if the comparison is true | no result, add info annotation | <li>- [x] </li> |
|`==` and `!=`  without `bool` modifier | `histogram`, `histogram` | treat both histograms as 0,  return nothing for `!=` and the first(?) histogram for `==`|check histograms for (in)equality, return the first `histogram` if the check is true | <li>- [x] </li> |
| `>`, `<`, `<=`, `>=`   without `bool` modifier| `histogram` on either or both sides | treat `histogram` as 0, return value depends on order | no result, add info annotation | <li>- [x] </li> |
|`count_values`| combination of `histogram` and `float` | treat `histogram` as 0 | use the string representation of histogram | <li>- [x] </li> |
|`topk`, `bottomk` | combination of `histogram` and `float` | treat `histogram` as 0 | ignore `histogram`, add info annotation | <li>- [x] </li> |
|`limitk`, `limit_ratio`| combination of `histogram` and `float` | treat `histogram` as 0 | add `histogram` to the result if applicable | <li>- [x] </li> |
|`sum_over_time`, `avg_over_time` | combination of `histogram` and `float` in one range | no result, add warn annotation | no result, add warn annotation (i.e. already correct, but we should add tests) | <li>- [x] </li> |
|`min_over_time`, `max_over_time`, `quantile_over_time`, `stddev_over_time`, `stdvar_over_time`, `mad_over_time`,`deriv`, `double_exponential_smoothing`, `predict_linear` | combination of `histogram` and `float` in one range | ignore histogram | ignore histograms, calculate results just for the floats in the range, add info annotation | <li>- [x] </li> |
|`min_over_time`, `max_over_time`, `quantile_over_time`, `stddev_over_time`, `stdvar_over_time`, `mad_over_time`,`deriv`, `double_exponential_smoothing`, `predict_linear` | only histograms in one range | ignore | ignore (i.e. already correct, but we should add tests) | <li>- [x] </li> |
|`idelta`, `irate`| the last two samples in range are both histograms | no result | calculate result corresponding to `delta`/`rate`, including annotation about applying `idelta` to counter histogram and `irate` to gauge histogram | <li>- [x] </li> |
|`idelta`, `irate`| the last two samples in range are a histogram and a float sample | ignore histogram | no result, add warn annotation | <li>- [x] </li> |
|`sort`, `sort_by_desc`, `scalar` | histogram | treat histogram as `0` float | ignore histogram silently (which might, in the case of `scalar`, result in fulfilling the condition of exactly one (float) sample in the input vector)  | <li>- [x] </li> |

**Comment by beorn7** [comment]
Related: #11973

**Comment by beorn7** [comment]
My thoughts:

I think the (float, histogram) args for arithmetic binary operators needs to be broken up. For example, `float + histogram` or `histogram - float` doesn't work, but `float * histogram` and `histogram / float` is just fine (but `float / histogram` is not).

For `min` and `max`, I think it's fine to just return nothing for aggregations of only histograms. The warning should only be issued if the aggregation tries to aggregate histograms _and_ floats into the same key. (For example, if `foo` is a vector that has both floats and histograms, and all the histograms have a `kind="histogram"` label and all the floats have a `kind="float"` label, then `min by (kind)(foo)` should not give a warning, but `min(foo)` should.)

I think the current behavior of `count` is fine. `count` just counts sample, and a float and a histogram should be counted the same. (I think that's even documented.)

`histogram / 0` should return `sum` and `count` __and the zero bucket__ "normally" (I assume in the way it is done already), but then all (normal) buckets should be removed (because all the non-represented empty buckets should be `NaN` now, which we cannot represent. So better have no buckets at all. (This case is anyway pathologic.)

**Comment by beorn7** [comment]
The basic idea is this: Where operations and aggregations make sense (between histogram and histogram, or between histogram and float), they happen. Where they don't make sense, the histogram is treated as if the vector element with the histogram did not exist. This is not an error, but we attach a "warn"-level annotation (or maybe "info"-level only?) if any vector element is label-matched with an ignored histogram. That implies no annotation is added if two ignored histograms are matched or if an ignored histogram is matched with a scalar (which is not label-matching).

The rationale here is that label matching might be complex, and if a label match leads to an impossible combination, it might be accidental and the user should know about it. In contrast to that, `2 / foo` essentially means "Divide 2 by all the float elements and ignore the histogram elements" and that's just fine and can deliberately be used for filtering. Similarly, `min(foo)` means "give me the minimum of all float elements and ignore histograms", which is also fine. However, if I write `foo + bar`, the user's idea is probably that the label-matched elements should also have the same type, and if a float gets matched with a histogram, it's probably an error. (And the "probably" makes me think we should perhaps use "info" level, which implies that false positives are possible, while "warn" level is meant for annotations that are always triggered by something that is "wrong" in some way and could be fixed somehow.)

(This is all just my stream of thoughts. Feel free to express other viewpoints.)

**Comment by NeerajGartia21** [comment]
> I think the (float, histogram) args for arithmetic binary operators needs to be broken up. For example, float + histogram or histogram - float doesn't work, but float * histogram and histogram / float is just fine (but float / histogram is not). histogram + histogram is fine, but histogram * histogram is not.

Yes, sorry, I missed this case. I'll break the operators further with different arguments and  update the table.

> For min and max, I think it's fine to just return nothing for aggregations of only histograms. The warning should only be issued if the aggregation tries to aggregate histograms and floats into the same key. (For example, if foo is a vector that has both floats and histograms, and all the histograms have a kind="histogram" label and all the floats have a kind="float" label, then min by (kind)(foo) should not give a warning (but the only element in the result vector has kind="float"), but min(foo) should give a warning.)

So the gist here is that after grouping the labels if the vector contains both `histogram` and `float` then we'll issue a warning otherwise we will return the `min` in case of `float` only vector and nothing in case of `histogram` only vector. Please correct me if I misunderstood something.

> I think the current behavior of count is fine. count just counts sample, and a float and a histogram should be counted the same. (I think that's even documented.)

yeah, my bad, I somehow got confused between `count` and `histogram_count`.

I'll make all these changes into the table.

**Comment by NeerajGartia21** [comment]
> The rationale here is that label matching might be complex, and if a label match leads to an impossible combination, it might be accidental and the user should know about it. In contrast to that, 2 / foo essentially means "Divide 2 by all the float elements and ignore the histogram elements" and that's just fine and can deliberately be used for filtering. However, if I write foo + bar, the user's idea is probably that the label-matched elements should also have the same type, and if a float gets matched with a histogram, it's probably an error. (And the "probably" makes me think we should perhaps use "info" level, which implies that false positives are possible, while "warn" level is meant for annotations that are always triggered by something that is "wrong" in some way and could be fixed somehow.)

In my opinion if the operation makes sense and something is returned from that operation but we have ignored histogram in the vector ( like `2/foo` ) then we can add an info level annotation stating that histograms are ignored in the operation and if the operation does not makes any sense then I think we should add a warn level annotation that this operation for the given arguments is not possible (or something like that). These are my thoughts pls correct me.

**Comment by beorn7** [comment]
> So the gist here is that after grouping the labels if the vector contains both histogram and float then we'll issue a warning otherwise we will return the min in case of float only vector and nothing in case of histogram only vector.

Yes, that's the idea (but the case of how to deal with histogram-only could still be debated, see your later comment).

**Comment by beorn7** [comment]
> In my opinion if the operation makes sense and something is returned from that operation but we have ignored histogram in the vector ( like 2/foo ) then we can add an info level annotation stating that histograms are ignored in the operation and if the operation does not makes any sense (like float+histogram ) then I think we should add a warn level annotation that this operation for the given arguments is not possible (or something like that). These are my thoughts pls correct me.

That's also an option. If I understand you correctly, you would add an info annotation to the `2 / foo` expression if `foo` contains both histograms and floats (so _something_ is returned) but if `foo` is a pure histogram vector, you would add a warning annotation (because nothing is returned). That might be hard to implement (you can only decide about the level of the annotation when looking at the whole vector), but maybe it's doable. I'm a bit more concerned about the change of behavior that a "float divided by histogram" operation becomes "more bad" if the histogram is not accompanied by a float in the same vector.

Another thought in this regard: A mix of floats and histograms in the same vector should be a very rare case. So maybe we don't need to think too much about treating that case specially or with concerns about spammy annotations. It shouldn't really happen in practice, and if it does, it's probably a faulty setup or a transition period and will be fixed soon. So maybe let's say for now, whenever we ignore a histogram in an operation because it doesn't make sense, we add an info-level annotation saying that an operation wasn't performed, even if the expression still yields a result from other vector elements. We can do this first, it's easy, and then iterate on it once we see how it plays out in practice.

**Comment by NeerajGartia21** [comment]
> if the histogram is not accompanied by a float in the same vector.

Does this mean a histogram and float sample in the same timeseries? Or do they belong to different timeseries, but appear together in the result after label matching?

> So maybe let's say for now, whenever we ignore a histogram in an operation because it doesn't make sense, we add an info-level annotation saying that an operation wasn't performed, even if the expression still yields a result from other vector elements. We can do this first, it's easy, and then iterate on it once we see how it plays out in practice.

Yeah, I can start working on this part and after that we'll see how it goes.

**Comment by beorn7** [comment]
I have updated the table above according to my last paragraph above.

**Comment by NeerajGartia21** [comment]
Thanks for updating the table Beorn. I was busy and couldn't update the table fully.

**Comment by NeerajGartia21** [comment]
Thanks, @beorn7, for updating the table. In the last few months, I wasn't feeling well and was diagnosed with an autoimmune disease. However, I am now fine and can continue working on this.

**Comment by beorn7** [comment]
All the best for your further recovery. I'll assign this to you then, but no pressure! If you feel overwhelmed, just let me know.

**Comment by NeerajGartia21** [comment]
Thanks @beorn7 for re-assigning it to me. I'll let you know if I face any problem!

**Comment by NeerajGartia21** [comment]
/ | float, histogram (in this order! histogram, float is fine) | returns float ignoring histogram | no result, add info annotation
-- | -- | -- | --

@beorn7 In this should we add `histogram`/`histogram` as well? Because currently it produces the results by dividing floats.

**Comment by beorn7** [comment]
Good point.

One might argue, since we allow `histogram / float`, yielding another histogram, we should also allow the inverse `histogram / histogram`, yielding a float. However, that would only work if the numbers in the one histograms are all related by the same factor to the numbers in the other histogram, and that's not generally the case.

So indeed, I think `histogram / histogram` should behave in the same way as `histogram * histogram`. I'll modify the table accordingly, but also happy to hear other thoughts.

**Comment by NeerajGartia21** [comment]
Yeah, `histogram / histogram` is not a general case so, I think we can return an info annotation with no results in that case. There is no point implementing this functionality IMO.

**Comment by charleskorn** [comment]
https://github.com/prometheus/prometheus/pull/14821 might also be relevant to this issue.

**Comment by beorn7** [comment]
#14821 raises two important questions (and also includes the required fixes and tests):

1. Unary `-` doesn't do anything, but should (of course) do the same as `-1 * histogram` or `0-histogram`.
2. Should a histogram ever have a negative count or negative bucket populations?

I think the answer for (1) is pretty clear: Yes, those things should be consistent, and we need to implement unary `-`.

(2) is trickier. While a negative count might seem nonsensical, I think within PromQL it's just a mathematical construct (a bit similar to a "weighted" histogram, where you get non-integer counts and bucket populations). For example, if you want to remove a set of observations represented by `histogram_1` from a (larger) `histogram_2`, you could write `histogram_2 - histogram_1`. But this should yield the same result as `histogram_2 + (-1 * histogram_1)` or `histogram_2 + (-histogram_1)`. The overall result should not have negative counts or bucket populations, but users can of course write nonsensical stuff like `histogram_1 - histogram_2`. But that's the way it is, we cannot "fix" things be simply inverting the sign of the count or of bucket populations. (Another similar thing is that you can write `http_requests_total * -5`, i.e. you create a "negative counter".)

So in summary, I think we have to allow "negative histograms" as something in PromQL, even if it might appear nonsensical in isolation.

**Comment by charleskorn** [comment]
Your reasoning in https://github.com/prometheus/prometheus/issues/13934#issuecomment-2328272160 makes sense to me - I'll update #14821 to match this.

**Comment by charleskorn** [comment]
What about comparison operations like `==` and `>`? At present, native histograms are treated as if they're 0s, which produces unexpected results.

For valid combinations (eg. `histogram == other_histogram`), I'd expect the result to be based on comparing the histograms, and for nonsensical combinations (eg. `histogram == 3` or `histogram > other_histogram`) I'd expect no result and an annotation saying that histograms can't be compared to scalars or that `>` isn't defined for histograms, for example.

**Comment by beorn7** [comment]
Thanks for the finding. I somehow thought we had comparison operators nailed already. Will add another line to our grand table above.

**Comment by beorn7** [comment]
Which is done now.

**Comment by beorn7** [comment]
With https://github.com/prometheus/prometheus/pull/14726 merged, we can probably start to add a "Done" column to the table above to track progress.

**Comment by NeerajGartia21** [comment]
I've added a status column to the table ! And I've started working on the remaining tasks.

**Comment by NeerajGartia21** [comment]
@beorn7 , In case of SUM we add a warn annotation which is `NewMixedFloatsHistogramsAggWarning` but for other aggregations we are adding info annotations. Isn't is quite inconsistent?

**Comment by beorn7** [comment]
I think it depends on the aggregation. If the mix between a histogram and a float makes the aggregation impossible, so that the result disappears completely (sum, avg), we should add a warn annotation. If the aggregation ignores histograms, but still yields a result (max, min, stddev, stdvar, bottomk, topk, quantile), we should add an info annotation.

(Aggregations not mentioned above should "just work" even with a mix of histograms and floats, but we have to check if that's really the case. `count_values` comes to mind as weird – how to put a histogram value into a label? In its string representation? Maybe that's the right thing to do…)

**Comment by NeerajGartia21** [comment]
But in case of Sum we can always ignore the histograms like we do in case of Add operator. Is there any specific reason to make it different from Add operator? 

> (Aggregations not mentioned above should "just work" even with a mix of histograms and floats, but we have to check if that's really the case. `count_values` comes to mind as weird – how to put a histogram value into a label? In its string representation? Maybe that's the right thing to do…)

Yeah I think for `count_values` we can use it's string representation.

**Comment by beorn7** [comment]
I don't think we are ignoring histograms in `sum` and `+`:

For the `sum` aggregation, we do not ignore histograms. For any result that would require the aggregation of histograms and floats, we return nothing (remove that group from the result), and attach a warning annotation.

For the `+` binop (I guess that's what you mean by "add operator"), we also remove results that try to add a histogram to a float, but since it is a binop, it's plausible that this could be used as an intended filtering, so we only attach an info annotation.

I'm not 100% confident that this is the best approach, but we discussed it before, and this is the outcome as I understood it:
- If a histogram vs. float collision removes a group from an aggregation, we annotate with warn level.
- If a histogram vs. float collision removes an element from the result of a binop, we annotate with info level.

If we have new evidence that this is not a good plan, we can revise it. But for now, I think the aggregation case is indeed much different from the binop case. Aggregation is "worse" indeed as you could really aggregate a lot of histograms, and a single float accidentally sneaking in will kill that whole aggregation group (thereby possible preventing your alert from firing). With a binop, the collisions could still be accidental, of course, but you are only ever matching two elements, so it's much easier to keep under control. It's also "normal" in binops that elements disappear that do not label-match, so one might argue the histogram/float property behaves just like a label here. Plus, as said above, even relying on the behavior on purpose (for filtering) is a much more plausible approach than it would be with aggregations.

**Comment by NeerajGartia21** [comment]
Got it ! Thanks for the explanation @beorn7 . 
Also apart from operators listed above I think we need to modify the behaviour of `atan` operator as well. Should I add it to the table if that is the case?

**Comment by beorn7** [comment]
Yes, the table is certainly not complete yet. Please amend as you see fit.

**Comment by krajorama** [comment]
I'd like to challenge the last line ">, <, <=, >= without bool modifier" . Especially because we have clamp function and also the proposed trim (>/) functions, I think it would make sense to interpret conditions like this (using "<" for example).

histogram < float  would mean _all_ observations are below the float value
float < histogram would mean the float is below _all_ observations in the histogram

histogram < histogram would return nothing and warning. There's probably a way to define this as well to mean more observations are higher in the right side - the trivial case being that _all_ observations are higher in the right side.

**Comment by beorn7** [comment]
More thoughts about @krajorama's idea [here](https://github.com/prometheus/prometheus/issues/14651#issuecomment-2485736158).

**Comment by NeerajGartia21** [comment]
After investing for the remaining operators here are my thoughts:
- For `count_values` we can represent histogram in their string format ( which is done in #15422 ).
- For `topk` and `bottomk` we should ignore the histograms and add info annotations for them.
- For `limitk` and `limit_ratio` we should not ignore the histograms and should be included in the results(if applicable).

Please share your thoughts. I'll update the table accordingly.

**Comment by beorn7** [comment]
Sounds about right. Note that https://github.com/prometheus/docs/pull/2539 should contain all the desired behavior.

**Comment by NeerajGartia21** [comment]
I think in [prometheus/docs#2539](https://github.com/prometheus/docs/pull/2539) it is already mentioned which can be found [here](https://github.com/prometheus/docs/blob/ac3275cb462aa9bd03c0894b0a4af1ec7acfe303/content/docs/specs/native_histograms.md#aggregation-operators).

**Comment by beorn7** [comment]
Yes. Just trying to say that my thoughts are expressed in that document. If my thoughts match your thoughts, you could just update it in the table. If you have any concerns, let's discuss here in more detail.

**Comment by NeerajGartia21** [comment]
I have reviewed a few simple functions, including 
`abs()`, `ceil()`, `clamp()`, `clamp_max()`, `clamp_min()`, `day_of_month()`, `day_of_week()`, `day_of_year()`, `days_in_month()`, `deriv()`, `exp()`, `floor()`, `hour()`, `ln()`, `log2()`, `log10()`, `minute()`, `month()`, `round()`, `scalar()`, `sgn()`, `sqrt()`, `time()`, `vector()`, and `year()`. 
These functions should only operate with floats and ignore histograms. However, do we need to include any annotations for them in cases where histograms are present?

**Comment by beorn7** [comment]
Quoting from https://github.com/prometheus/docs/pull/2539 : "All remaining functions not mentioned in this section do _not_ work with native histograms. Histogram elements in the input vector are silently ignored."

So yes, the intended behavior is to not add any annotations in those cases.

**Comment by beorn7** [comment]
`idelta` and `irate` still need an implementation for histograms, don't they?

**Comment by NeerajGartia21** [comment]
Yes, `idelta` and `irate` are not implemented for histogram yet. I'll try to do it ASAP.

**Comment by beorn7** [comment]
Just noticing that `deriv()`, `double_exponential_smoothing()`, and `predict_linear()`, should probably behave like the `..._over_time` functions that ignore native histograms, i.e. silently ignore pure native histogram samples, and filtering out the histograms in mixed ranges, flagging it by an info annotation.

**Comment by beorn7** [comment]
More functions that should ignore histograms, but (presumably) still treat them as zero:
- `sort` and `sort_by_desc` (but note that `sort_by_label` and `sort_by_label_desc` are fine).
- `scalar`

Sorry for just dumping all of this. I'm stumbling on these while updating the PromQL documentation. We should add them to the table in a clean fashion.

**Comment by beorn7** [comment]
> Just noticing that `deriv()`, `double_exponential_smoothing()`, and `predict_linear()`, should probably behave like the `..._over_time` functions that ignore native histograms, i.e. silently ignore pure native histogram samples, and filtering out the histograms in mixed ranges, flagging it by an info annotation.

This wasn't even in the spec. Added it, see https://github.com/prometheus/docs/pull/2557 .

**Comment by beorn7** [comment]
I think we are done here! 🎉 

Special thanks to @NeerajGartia21 who dutifully hunted down all of these.

**Comment by NeerajGartia21** [comment]
Enjoyed working on this. Thanks, everyone! 🙌
